internal/controller/role/role.go

```diff
@@ -0,0 +1,292 @@
+// is called.
+type connector struct {
+	kube         client.Client
+	usage        resource.Tracker
+	newServiceFn func(creds []byte) (interface{}, error)
+	logger       logging.Logger
+}
+
+// Connect typically produces an ExternalClient by:
+// 1. Tracking that the managed resource is using a ProviderConfig.
+// 2. Getting the managed resource's ProviderConfig.
+// 3. Getting the credentials specified by the ProviderConfig.
+// 4. Using the credentials to form a client.
+func (c *connector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	cr, ok := mg.(*v1alpha1.Role)
+	if !ok {
+		return nil, errors.New(errNotRole)
+	}
+
+	vaultClient, err := clients.NewVaultClient(ctx, c.kube, cr)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewExternalClient)
+	}
+
+	return &external{
+		client: vaultClient,
+		logger: c.logger,
+	}, nil
+}
+
+// An ExternalClient observes, then either creates, updates, or deletes an
+// external resource to ensure it reflects the managed resource's desired state.
+type external struct {
+	// A 'client' used to connect to the external resource API. In practice this
+	// would be something like an AWS SDK client.
+	client clients.VaultClient
+
+	logger logging.Logger
+}
+
+func (c *external) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	role, ok := mg.(*v1alpha1.Role)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotRole)
+	}
+
+	name := role.Name
+	authBackend := role.Spec.ForProvider.Backend
+
+	path := authBackend + "/roles/" + name
+
+	c.logger.Debug("[DEBUG] Reading role from %q", path)
+	_, err := c.client.Logical().Read(path)
+	if err != nil {
+		return managed.ExternalObservation{}, fmt.Errorf("error reading role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read role from %q", path)
+
+	return managed.ExternalObservation{
+		// Return false when the external resource does not exist. This lets
+		// the managed resource reconciler know that it needs to call Create to
+		// (re)create the resource, or that it has successfully been deleted.
+		ResourceExists: true,
+
+		// Return false when the external resource exists, but it not up to date
+		// with the desired managed resource state. This lets the managed
+		// resource reconciler know that it needs to call Update.
+		ResourceUpToDate: true,
+
+		// Return any details that may be required to connect to the external
+		// resource. These will be stored as the connection secret.
+		ConnectionDetails: managed.ConnectionDetails{},
+	}, nil
+}
+
+// Create an AWS Secret Backend Role
+func (c *external) Create(ctx context.Context, mg resource.Managed) (managed.ExternalCreation, error) {
+	role, ok := mg.(*v1alpha1.Role)
+	if !ok {
+		return managed.ExternalCreation{}, errors.New(errNotRole)
+	}
+
+	_, err := c.writeRole(role)
+	if err != nil {
+		return managed.ExternalCreation{}, fmt.Errorf("error creating role")
```

Comment:
Roberto has implemented some default error types. I think we should leverage it and wrap the error with that one.